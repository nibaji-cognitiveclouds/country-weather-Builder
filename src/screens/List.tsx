/** @format */

import axios from "axios";
import React, { FC, useEffect, useState } from "react";
import {
	FlatList,
	SafeAreaView,
	View,
	Text,
	Image,
	TouchableOpacity,
	ActivityIndicator,
	Modal,
	Button,
} from "react-native";
import { screens } from "../styles/screens";
import { countriesType, weatherType } from "../types/responses";
import { listScreenPropsType } from "../types/screens";

const List: FC<listScreenPropsType> = (props) => {
	const [countryData, setCountryData] = useState<countriesType[]>([]);
	const [loading, setLoading] = useState<boolean>(false);
	const [error, setError] = useState<boolean>(false);
	const [showModal, setShowModal] = useState<boolean>(false);

	const [weatherData, setWeatherData] = useState<weatherType>();

	useEffect(() => {
		getCountryData();
	}, []);

	function getCountryData() {
		setLoading(true);
		axios
			.get(`https://restcountries.com/v3/name/${props.route.params.name}`)
			.then((res) => {
				setCountryData(res.data);
			})
			.catch(() => {
				setError(true);
			})
			.finally(() => {
				setLoading(false);
			});
	}

	function handlPress(capital: string) {
		axios
			.get(
				`http://api.weatherstack.com/current?access_key=9cfe09f6522f28c80de18d51c2880bf4&query=${capital}`
			)
			.then((res) => setWeatherData(res.data))
			.catch(() => {
				setError(true);
			})
			.finally(() => {
				!error && setShowModal(true);
			});
	}

	return (
		<SafeAreaView testID="lists" style={screens.container}>
			<Modal visible={showModal} animationType={"slide"}>
				<View style={screens.modal}>
					<Image
						source={{ uri: `${weatherData?.current?.weather_icons[0]}` }}
						style={{ height: 100, width: 100 }}
					/>
					<Text>Temperature : {weatherData?.current?.temperature} C</Text>
					<Text>Wind Speed : {weatherData?.current?.wind_speed}</Text>
					<Text>Precipitation : {weatherData?.current?.precip}</Text>
					<Button
						title={"Ok"}
						onPress={() => {
							setShowModal(false);
						}}
					/>
				</View>
			</Modal>

			<FlatList
				data={countryData}
				renderItem={({ item }) => (
					<TouchableOpacity
						onPress={() => {
							handlPress(item?.capital[0]);
						}}
					>
						<View style={screens.item}>
							<View>
								<Text>Name : {item?.name.common}</Text>
								<Text>Capital : {item?.capital}</Text>
								<Text>Population : {item?.population}</Text>
								<Text>Latitude : {item?.latlng[0]}</Text>
								<Text>Longitude : {item?.latlng[1]}</Text>
							</View>
							<View>
								<Image
									source={{ uri: item?.flags[1] }}
									style={{ height: 100, width: 100 }}
								/>
							</View>
						</View>
					</TouchableOpacity>
				)}
				ListEmptyComponent={() => {
					return (
						<View>
							{loading ? (
								<ActivityIndicator color={"blue"} />
							) : !error ? (
								<Text>Nothing found</Text>
							) : (
								<Text>Something went wrong!</Text>
							)}
						</View>
					);
				}}
			/>
		</SafeAreaView>
	);
};

export default List;
