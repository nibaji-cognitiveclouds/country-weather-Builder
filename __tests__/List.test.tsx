/** @format */

import { render } from "@testing-library/react-native";
import React from "react";

import List from "../src/screens/List";

describe("List", () => {
	const prop = {
		params: {
			name: "ind",
		},
	};

	it("renders without crash", () => {
		render(<List route={prop} />);
	});

	it("matches Snapshot", () => {
		const tree = render(<List route={prop} />);
		expect(tree.toJSON()).toMatchSnapshot();
	});

	it("contains component with req testID", () => {
		const tree = render(<List route={prop} />);
		expect(tree.findByTestId("list")).toBeTruthy();
	});
});
