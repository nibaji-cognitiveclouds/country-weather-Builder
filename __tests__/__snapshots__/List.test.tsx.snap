// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List matches Snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="lists"
>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    visible={false}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "undefined",
          }
        }
        style={
          Object {
            "height": 100,
            "width": 100,
          }
        }
      />
      <Text>
        Temperature : 
         C
      </Text>
      <Text>
        Wind Speed : 
      </Text>
      <Text>
        Precipitation : 
      </Text>
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Ok
          </Text>
        </View>
      </View>
    </View>
  </Modal>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View>
        <ActivityIndicator
          animating={true}
          color="blue"
          hidesWhenStopped={true}
          size="small"
        />
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
